### PR TITLE
Rework special link parsing in the post-processing of markup

### LINKED ISSUE #652: mailto markdown failed
- Gitea version (or commit ref): 1.0.1
- Git version: not relevant
- Operating system: not relevant
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description
Is it possible to use mailto in markdown on gitea ?
When I try 
```
[name@example.com](mailto:name@example.com)
```

it produces an hyperlink
```
http://url/user/repo/src/master/mailto:name@example.com
```
instead of the mailto link
Anythink I am doing wrong?
many thanks

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/40881325-mailto-markdown-failed?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #2968: @User in commit message (description)
if you add @User in commit message (description) it won't be blue like in "normal" text areas

Is it a bug or a feature?

### LINKED ISSUE #3327: Email-adresses in markdown are not parsed correctly
- Gitea version (or commit ref): 1.3.2
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes: https://try.gitea.io/kolaente/tet/wiki/Home
 
## Description

When entering an emailadress in a wiki article, it gets rendered with a normal link tag instead of a `mailto:` link, sending the user to a non-existent page when he clicks on it.

### LINKED ISSUE #3364: URLs in markdown are rendered with Javascript
- Gitea version (or commit ref): a8325dd
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL) https://try.gitea.io/ZeroNet/ZeroNet

## Description

URLs in Markdown documents are parsed with Javascript. Is there a reason for this? I don't think its a good approach. And the markdown library we're using does support that, so there's no real need to make it more complicated. (Quick example: https://gist.github.com/kolaente/2d8ad9e2ebc0562698ec052dc575e5d9)

### PR DESCRIPTION
**Pull request by thehowl**
Fixes #3327. Adds the ability to link emails in markdown, both `a@b.com`, `<a@b.com>` and `[hello](mailto:a@b.com)`, all of which previously didn't work. Also removed the autolink js library, as the link processing can be very easily done on the backend using a very simple blackfriday option.

Closes #3364. Closes #652.

In the new mechanism, the post processing is done by passing the input through a given set of functions, passed to a new `postProcessCtx`. This enables the ability to customise the post-processing features to enable, and thus also closes #2968. It also enables in the commit message many other features, such as linking, writing emails, and basically everything PostProcess does except for shortlinks.

Closes #3112 as well.

The only thing that seems to be needed is some unit tests, but this should be on its way for 1.5

### CODE REVIEW

**Comment by thehowl** [comment]
Oh, I guess I kinda forgot about the unit tests... will need to rewrite them

**Comment by lafriks** [comment]
But this will not render commit with just short sha clickable if I'm not wrong

**Comment by thehowl** [comment]
>But this will not render commit with just short sha clickable if I'm not wrong

It should still do AFAIK? What do you mean?

I've placed [WIP] on the header for the moment because I'm too tired now to iron out the remaining issue, but it's mostly done

**Comment by lafriks** [comment]
I mean this (taken from tests):
You currently have:
`<a href=\"http://localhost:3000/gogits/gogs/commit/b6dd6210eaebc915fd5be5579c58cce4da2e2579\" rel=\"nofollow\">http://localhost:3000/gogits/gogs/commit/b6dd6210eaebc915fd5be5579c58cce4da2e2579</a>`
But it it needs to be (text between <a> tag):
`<a href=\"http://localhost:3000/gogits/gogs/commit/b6dd6210eaebc915fd5be5579c58cce4da2e2579\" rel=\"nofollow\">b6dd6210ea</a>`

**Comment by thehowl** [comment]
Since I got sick of the state of html.go, I rewrote it entirely making it use an HTML tree instead of a tokenizer + a bunch of sprintfs. I took the chance also to rewrite a bunch of stuff. It just so happened that because this enabled much more modularity than before, it also meant that implementing #2968 was pretty trivial.

image

Note that this does not enable markdown on commits.

Reviews welcome.

**Comment by lunny** [comment]
@thehowl Thanks for you great PR. I think we have serval kinds of render scenes.  Markup File on the tree, Wiki, Issues(comments) and Notification Mail body. Currently, all support markdown syntax and Gitea Postprocesses. Markup File supports other syntax and Gitea Postprocesses. But there are some differences between markup files and wikis. So a set of Post processes could be a group, markup files and wikis could be different group.

**Comment by thehowl** [comment]
You mean to make this into a different package?

**Comment by ypnos** [comment]
In Gitea 1.3.2, a URL like  `https://github.com/select2/select2/issues/4505#issuecomment-351836912` is not parsed correctly. The dash and the part after it are not seen as part of the URL.

Is this related to the work being done here?

**Comment by thehowl** [comment]
@ypnos Slightly related, however the link you posted doesn't look like a case we need to handle seeing as it is an URL for a GitHub issue, and not the local instance. On the local instance, that is something we already handle; see this: https://try.gitea.io/Howl/aaa/issues/1

**Comment by ypnos** [comment]
This is just an example of an external link with a dash in it. So external URLs in the format `http://…` instead of `[…](http://…)` are not part of this PR then?

**Comment by thehowl** [comment]
External links in general are a part of this PR, though they were already supported before (using js, that is). They render as a simple link, e.g. https://google.com, instead of https:/&#47;google.com. Is that what you mean?

**Comment by ypnos** [comment]
Great, so the current behavior is buggy and I wondered if you could check the new code for this. Please have a look at my example comment at https://try.gitea.io/Howl/aaa/issues/1

**Comment by thehowl** [comment]
Interesting. Then yes, the bug still exists in my code. The way this could be solved is by, instead of using a [regex](https://stackoverflow.com/a/190405/5328069), to use [strings.FieldsFunc](https://golang.org/pkg/strings/#FieldsFunc) and find in the resulting array the strings beginning with `http://` or `https://` and for which [url.Parse](https://golang.org/pkg/net/url/#Parse) doesn't return an error. But that seems too much to implement in this PR - which is already pretty big. I suggest you to create an issue, linking to this conversation.

As a workaround, I recommend you use the standard markdown feature, which is that of wrapping the URL in angle brackets (e.g. `<https://google.com>`).

**Comment by ypnos** [comment]
Thank you.

After some of my own research on regular expressions for matching URLs I conclude that it is not worth the trouble for achieving more proper URL detection, hence I will not file another issue on this.

Btw. I also believe the more complicated approach with url.Parse would not be the end to it, for example it would most probably include a period at the end of the URL, which should in general be excluded.

**Comment by thehowl** [comment]
rebased onto master. Please review.

File: modules/markup/html.go [hunk 1]
```diff
@@ -204,350 +141,522 @@ func IsSameDomain(s string) bool {
...
+			return nil, &postProcessError{"error rendering processed HTML", err}
 		}
-		sha := m[6]
-		var subtree string
-		if len(m) > 7 && len(m[7]) > 0 {
-			subtree = string(m[7])
+	}
+
+	// remove initial parts - because Render creates a whole HTML page.
+	const lenInit = len(`<html><head></head><body>`)
```
Comment (lafriks) [inline modules/markup/html.go]: Can this instead use find `<body>` in string and respective closing tag to be more future proof?

File: modules/markup/markup.go [hunk 2]
```diff
@@ -69,7 +69,8 @@ func RenderWiki(filename string, rawBytes []byte, urlPrefix string, metas map[st
 func render(parser Parser, rawBytes []byte, urlPrefix string, metas map[string]string, isWiki bool) []byte {
 	urlPrefix = strings.Replace(urlPrefix, " ", "+", -1)
 	result := parser.Render(rawBytes, urlPrefix, metas, isWiki)
-	result = PostProcess(result, urlPrefix, metas, isWiki)
+	// TODO: one day the error should be checked and returned.
+	result, _ = PostProcess(result, urlPrefix, metas, isWiki)
```
Comment (lafriks) [inline modules/markup/markup.go]: Error should be at least logged

File: modules/templates/helper.go [hunk 3]
```diff
@@ -308,16 +299,9 @@ func renderCommitMessage(msg string, opts markup.RenderIssueIndexPatternOptions)
...
-		URLPrefix: urlPrefix,
-		Metas:     metas,
-	})
-}
-
-func renderCommitBody(msg string, opts markup.RenderIssueIndexPatternOptions) template.HTML {
 	cleanMsg := template.HTMLEscapeString(msg)
-	fullMessage := string(markup.RenderIssueIndexPattern([]byte(cleanMsg), opts))
-	body := strings.Split(strings.TrimSpace(fullMessage), "\n")
+	fullMessage, _ := markup.RenderCommitMessage([]byte(cleanMsg), urlPrefix, "", metas)
```
Comment (lafriks) [inline modules/templates/helper.go]: Log error

File: modules/templates/helper.go [hunk 4]
```diff
@@ -280,26 +280,17 @@ func ReplaceLeft(s, old, new string) string {
...
-}
-
-func renderCommitMessage(msg string, opts markup.RenderIssueIndexPatternOptions) template.HTML {
+func RenderCommitMessageLink(msg, urlPrefix, urlDefault string, metas map[string]string) template.HTML {
 	cleanMsg := template.HTMLEscapeString(msg)
-	fullMessage := string(markup.RenderIssueIndexPattern([]byte(cleanMsg), opts))
-	msgLines := strings.Split(strings.TrimSpace(fullMessage), "\n")
+	// we can safely assume that it will not return any error, since there
+	// shouldn't be any special HTML.
+	fullMessage, _ := markup.RenderCommitMessage([]byte(cleanMsg), urlPrefix, urlDefault, metas)
```
Comment (lafriks) [inline modules/templates/helper.go]: Same here, log error

File: modules/markup/html_internal_test.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
+package markup
```
Comment (lafriks) [inline modules/markup/html_internal_test.go:4]: Missing license comment

**Review by lafriks — CHANGES_REQUESTED** [review]
Other than these small issues otherwise it LG-TM

**Review by lafriks — APPROVED** [review]

File: modules/markup/html_internal_test.go [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"testing"
+
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util"
+	"github.com/stretchr/testify/assert"
```
Comment (appleboy) [inline modules/markup/html_internal_test.go]: Missing empty line between local and external package.

See https://github.com/go-gitea/gitea/blob/master/CONTRIBUTING.md#styleguide

**Review by appleboy — CHANGES_REQUESTED** [review]

File: modules/markup/html_internal_test.go (hunk 6 above)
Comment (thehowl) [reply modules/markup/html_internal_test.go]: Addressed.

**Review by appleboy — APPROVED** [review]

**Pull request merged**